packages/core/src/wallet/metamaskBridge.ts

```diff
@@ -0,0 +1,203 @@
+import { loadRemote, init } from "@module-federation/runtime"
+import type {
+  IStarknetWindowObject,
+  RpcMessage,
+  StarknetWindowObject,
+  WalletEvents,
+} from "../StarknetWindowObject"
+import wallets, { WalletProvider } from "../discovery"
+
+interface MetaMaskProvider {
+  isMetaMask: boolean
+  request(options: { method: string }): Promise<void>
+}
+
+declare global {
+  interface Window {
+    ethereum?: MetaMaskProvider
+  }
+}
+
+function detectMetaMaskProvider({ timeout = 3000 } = {}) {
+  let handled = false
+  return new Promise<MetaMaskProvider | null>((resolve) => {
+    if (window.ethereum) {
+      handleEthereum()
+    } else {
+      window.addEventListener("ethereum#initialized", handleEthereum, {
+        once: true,
+      })
+      setTimeout(() => {
+        handleEthereum()
+      }, timeout)
+    }
+    function handleEthereum() {
+      if (handled) {
+        return
+      }
+      handled = true
+      window.removeEventListener("ethereum#initialized", handleEthereum)
+      const { ethereum } = window
+      if (ethereum && ethereum.isMetaMask) {
+        resolve(ethereum)
+      } else {
+        resolve(null)
+      }
+    }
+  })
+}
+
+async function waitForMetaMaskProvider(
+  options: { timeout?: number; retries?: number } = {},
+): Promise<MetaMaskProvider | null> {
+  const { timeout = 3000, retries = 0 } = options
+
+  let provider: MetaMaskProvider | null = null
+  try {
+    provider = await detectMetaMaskProvider({ timeout })
+  } catch {
+    // Silent error - do nothing
+  }
+
+  if (provider) {
+    return provider
+  }
+
+  if (retries === 0) {
+    return null
+  }
+
+  provider = await waitForMetaMaskProvider({ timeout, retries: retries - 1 })
+  return provider
+}
+
+async function hasSnapSupport(provider: MetaMaskProvider) {
+  try {
+    await provider.request({ method: "wallet_getSnaps" })
```

Comment:
use non-api call to check for **starknet** support (not just and snaps?)

Reply:
hi @avimak the life cycle of a snap is working like that way

1. you detect where you have a install metamask
2. you detect where you have an correct version of metamask -> which using similar method as above
3. then you ask metamask to install the starknet snap -> that should be trigger by enable() / or any similar method

so you can only cehck if the metamask support starknet by first checking if it has supported snap

In addtional, metamask does not inject any extra method to the provider (window.ethereum) to verify if it is support snap, it means, you must using request method to determined 

does it help to clarify your questions?

Reply:
@avimak 
After consulting with @stanleyyconsensys we decided to remove that check.